framework/sandstone.cpp

```diff
@@ -3208,418 +2671,12 @@ int main(int argc, char **argv)
-            break;
-
-        case max_test_loop_count_option:
-            sApp->max_test_loop_count = ParseIntArgument<>{"--max-test-loop-count"}();
-            if (sApp->max_test_loop_count == 0)
-                    sApp->max_test_loop_count = std::numeric_limits<int>::max();
-            break;
-
-            /* deprecated options */
-        case longer_runtime_option:
-        case max_concurrent_threads_option:
-        case mem_sample_time_option:
-        case mem_samples_per_log_option:
-        case no_mem_sampling_option:
-        case no_triage_option:
-        case schedule_by_option:
-        case shortened_runtime_option:
-        case triage_option:
-        case weighted_testrun_option:
-            warn_deprecated_opt(long_options[coptind].name);
-            break;
-
-        case 0:
-            /* long option setting a value */
-            continue;
-        default:
-            suggest_help(argv);
-            return EX_USAGE;
-        }
-    }
-
-    if (SandstoneConfig::RestrictedCommandLine) {
-        // Default options for the simplified OpenDCDiag cmdline
-        static struct option restricted_long_options[] = {
-            { "help", no_argument, nullptr, 'h' },
-            { "query", no_argument, nullptr, 'q' },
-            { "service", no_argument, nullptr, 's' },
-            { "version", no_argument, nullptr, version_option },
-            { nullptr, 0, nullptr, 0 }
-        };
-
-        while ((opt = simple_getopt(argc, argv, restricted_long_options)) != -1) {
-            switch (opt) {
-            case 'q':
-                // ### FIXME
-                fprintf(stderr, "%s: --query not implemented yet\n", argv[0]);
-                abort();
-            case 's':
-                // keep in sync above
-                sApp->endtime = MonotonicTimePoint::max();
-                sApp->service_background_scan = true;
-                break;
-            case version_option:
-                logging_print_version();
-                return EXIT_SUCCESS;
-
-            case 'h':
-                usage(argv);
-                return opt == 'h' ? EXIT_SUCCESS : EX_USAGE;
-            default:
-                suggest_help(argv);
-                return EX_USAGE;
-            }
-        }
-
-        if (SandstoneConfig::NoLogging) {
-            sApp->shmem->output_format = SandstoneApplication::OutputFormat::no_output;
-        } else  {
-            sApp->shmem->verbosity = 1;
-        }
-
-        sApp->delay_between_tests = 50ms;
-        sApp->thermal_throttle_temp = INT_MIN;
-        fatal_errors = true;
-        builtin_test_list_name = "auto";
-
-        static_assert(!SandstoneConfig::RestrictedCommandLine || SandstoneConfig::HasBuiltinTestList,
-                "Restricted command-line build must have a built-in test list");
+    ParsedOpts opts;
+    auto ret = parse_cmdline(argc, argv, sApp, opts);
+    if (ret == EXIT_SUCCESS || ret == EX_USAGE) {
```

Comment:
I think that now we have a chance to improve on separation of parsing vs. acting. What if we introduced "commands" concept that would indicate an action that is to be taken (e.g. "list tests") as opposed to doing it in the place where we're parsing them. That is, after `parse_cmdline()` returns, `ParsedOpts` would contain specific single command that the subsequent code needs to execute and then `switch (opts.command) { case HELP: ...; break; case LIST_TESTS: ...; break; etc.}` with the default command being `RUN` which would actually proceed with the execution.

The same applies to changes to the configuration that need to be taken (e.g. applying different topology/cpuset). Ideally, none of that should be happening inside `sandstone_opts.cpp`, but here, in `main()`...

Reply:
this particular PR is intended to only separate the code. I have a followup which refactors the commandline parsing completely, and also does this 'parsing vs acting' separation. So if you don't mind, I would not perform any major changes to the parsing itself in this PR.

Reply:
I do mind. The reason I am mentioning this is that you change API and passing around sApp to accomplish the separation. What I am asking for is `enum Command` and a field in the options struct, then you can keep the action code in `sandstone.cpp`.